### PR TITLE
Spring Kafka Project

### PR DESCRIPTION
**Pull request by obedjamir**
first commit

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+			<artifactId>spring-kafka-test</artifactId>
+			<version>${spring-kafka-test.version}</version>
+		</dependency>
+		<dependency>
+			<groupId>com.opencsv</groupId>
+			<artifactId>opencsv</artifactId>
+			<version>${opencsv.version}</version>
+		</dependency>
+		<dependency>
+			<groupId>ch.qos.logback</groupId>
```
Comment (PinakiGhosh) [inline pom.xml]: This is not required

File: service/src/main/java/com/bankbazaar/kafka/service/controller/KafkaController.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.validation.Valid;
+
+@RestController
+public class KafkaController {
+    private final KafkaProducer producer;
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/kafka/service/controller/KafkaController.java:19]: Autowire the KafkaProducer

File: pom.xml [hunk 3]
```diff
@@ -49,31 +47,18 @@
...
-			<groupId>ch.qos.logback</groupId>
-			<artifactId>logback-classic</artifactId>
-			<version>${logback-classic.version}</version>
-		</dependency>
 		<dependency>
 			<groupId>org.springframework.boot</groupId>
 			<artifactId>spring-boot-starter-validation</artifactId>
 			<version> ${spring-boot-starter-validation.version}</version>
 		</dependency>
+		<dependency>
```
Comment (PinakiGhosh) [inline pom.xml]: Move all dependencies to the service pom

File: service/src/main/java/com/bankbazaar/kafka/service/consumer/KafkaConsumer.java [hunk 4]
```diff
@@ -13,7 +13,7 @@
 @Slf4j
 @Service
 public class KafkaConsumer {
-    @KafkaListener(topics = "create_CSV")
+    @KafkaListener(topics = "create_CSV", groupId = "Partner_Integration_Kafka_Group")
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/kafka/service/consumer/KafkaConsumer.java]: GroupId should be something related to your project. Also keep all configs like this in application.properties

File: service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import java.io.FileReader;
+import java.io.IOException;
+
+import static org.awaitility.Awaitility.await;
+import static org.junit.jupiter.api.Assertions.*;
+@SpringBootTest
+class KafkaControllerTest {
+
+    @Autowired
+    private KafkaController kafkaController;
```
Comment (PinakiGhosh) [inline service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java:37]: Use MockMvc how we used in the previous project. Not like this

File: service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        String[] nextRecord;
+        nextRecord = csvReader.readNext();
+
+        assertArrayEquals(nextRecord, dataDto.getHeaders().split(","));
+
+        int i=0;
+        while ((nextRecord = csvReader.readNext()) != null)
+        {
+            assertArrayEquals(nextRecord, dataDto.getData()[i].split("\\|"));
+            i++;
```
Comment (PinakiGhosh) [inline service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java:110]: Also assert on the number of lines

Comment (obedjamir) [reply service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java:110]: this has been implemented

File: service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java (hunk 5 above)
Comment (obedjamir) [reply service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java:37]: Mockmvc implemented for testing

File: service/src/main/java/com/bankbazaar/kafka/service/consumer/KafkaConsumer.java (hunk 4 above)
Comment (obedjamir) [reply service/src/main/java/com/bankbazaar/kafka/service/consumer/KafkaConsumer.java]: group id changed

File: pom.xml (hunk 3 above)
Comment (obedjamir) [reply pom.xml]: dependencies moved to service pom

File: core/src/main/java/com/bankbazaar/kafka/core/model/Data.java [hunk 7]
```diff
@@ -8,6 +8,7 @@
 @NoArgsConstructor
 @AllArgsConstructor
 public class Data {
+    private Long id;
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/kafka/core/model/Data.java:11]: Why is an id required?

File: dto/src/main/java/com/bankbazaar/kafka/dto/model/FileStatusDto.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.NoArgsConstructor;
+
+import javax.validation.constraints.NotNull;
+
+@lombok.Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class FileStatusDto {
```
Comment (PinakiGhosh) [inline dto/src/main/java/com/bankbazaar/kafka/dto/model/FileStatusDto.java]: No need for a FileStatusDto

File: service/pom.xml [hunk 9]
```diff
@@ -13,10 +13,25 @@
...
         <maven.compiler.source>11</maven.compiler.source>
         <maven.compiler.target>11</maven.compiler.target>
-        <mapstruct.version>1.5.0.RC1</mapstruct.version>
         <spring-kafka-test.version>2.8.5</spring-kafka-test.version>
         <opencsv.version>5.6</opencsv.version>
+        <spring-kafka.version>2.8.5</spring-kafka.version>
+        <kafka-streams.version>3.1.0</kafka-streams.version>
+        <awaitility.version>4.2.0</awaitility.version>
+        <spring.cloud.version>2021.0.2</spring.cloud.version>
+        <property name="hibernate.enable_lazy_load_no_trans" value="true"/>
```
Comment (PinakiGhosh) [inline service/pom.xml]: Why is this required?

File: service/src/main/java/com/bankbazaar/kafka/service/consumer/KafkaConsumer.java [hunk 10]
```diff
@@ -1,32 +1,72 @@
...
 
+import java.io.File;
 import java.io.FileWriter;
 import java.io.IOException;
-import java.io.Writer;
+import java.util.function.Consumer;
 
 @Slf4j
 @Service
 public class KafkaConsumer {
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/kafka/service/consumer/KafkaConsumer.java]: Rename to KafkaStreamsConfig. This class no longer is a single kafka consumer

File: service/src/main/java/com/bankbazaar/kafka/service/producer/KafkaProducer.java [hunk 11]
```diff
@@ -1,25 +1,41 @@
...
+    @Autowired
+    private FileStatusService fileStatusService;
 
     @Autowired
     private DataMapper mapper;
 
-    public void sendData(DataDto data)
+    public Response sendData(DataDto data)
     {
+        FileStatusDto fileData = new FileStatusDto();
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/kafka/service/producer/KafkaProducer.java]: Dtos are used in a scenario where we want to send data out of the application to a client. Its not required when we are calling other methods in the same service

File: service/src/main/java/com/bankbazaar/kafka/service/producer/KafkaProducer.java [hunk 12]
```diff
@@ -1,25 +1,41 @@
...
+        fileData.setFileName(data.getFileName());
+        fileData.setStatus("NEW");
+        FileStatusDto response = fileStatusService.insert(fileData);
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/kafka/service/producer/KafkaProducer.java]: This method should be like this
1. get Data object
2. Pass this object to the FileStatusService
3. FileStatusService should add the new entry and send back the entity
4. Push the request to the Kafka producer for processing
5. Create Response from the entity returned in step #3 and send it back

**Review by PinakiGhosh — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/bankbazaar/kafka/core/model/Data.java (hunk 7 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/kafka/core/model/Data.java:11]: I included id so that i can update the status in db after each process.

File: service/src/main/java/com/bankbazaar/kafka/service/producer/KafkaProducer.java (hunk 12 above)
Comment (obedjamir) [reply service/src/main/java/com/bankbazaar/kafka/service/producer/KafkaProducer.java]: This has been implemented

File: service/src/main/java/com/bankbazaar/kafka/service/producer/KafkaProducer.java (hunk 11 above)
Comment (obedjamir) [reply service/src/main/java/com/bankbazaar/kafka/service/producer/KafkaProducer.java]: FileStatusDTo removed

File: service/src/main/java/com/bankbazaar/kafka/service/consumer/KafkaConsumer.java (hunk 10 above)
Comment (obedjamir) [reply service/src/main/java/com/bankbazaar/kafka/service/consumer/KafkaConsumer.java]: all KafkaStream bean moved to KafkaStreamsConfig

File: service/pom.xml (hunk 9 above)
Comment (obedjamir) [reply service/pom.xml]: I forgot to remove this. This is not a required line.

File: dto/src/main/java/com/bankbazaar/kafka/dto/model/FileStatusDto.java (hunk 8 above)
Comment (obedjamir) [reply dto/src/main/java/com/bankbazaar/kafka/dto/model/FileStatusDto.java]: FileStatusDto removed

File: service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+                                    FileStatusEntity fileData = new FileStatusEntity();
+                                    fileData.setId(value.getId());
+                                    fileData.setStatus(Status.IN_PROGRESS);
+                                    fileStatusService.update(fileData);
+                                    ClassLoader classLoader = getClass().getClassLoader();
+                                    File file = new File(classLoader.getResource(".").getFile() + value.getFileName());
+                                    if (file.exists()) {
+                                        fileData.setStatus(Status.FAILURE);
+                                        fileStatusService.update(fileData);
+                                        return false;
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java]: This method is supposed to return of type Data and you are returning a boolean.

File: service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+     * Set status to SUCCESS if file IS created successfully.
+     * Set status to ERROR if exception occurs during file creation.
+     * Retry a maximum of 1 time if exception occurs;
+     */
+    @Bean
+    public Consumer<KStream<String,Data>> consumer(@Qualifier("streamRetryTemplate") RetryTemplate retryTemplate)
+    {
+        return kStream -> kStream.foreach((key, value) ->
+                {
+                    try {
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java]: The Retry template + context itself is like a try catch. No need to add it explicitly

File: service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+                                        File file = new File(classLoader.getResource(".").getFile() + value.getFileName());
+                                        FileWriter fileWriter = new FileWriter(file,false);
+                                        CSVWriter CsvWriter = new CSVWriter(fileWriter);
+                                        CsvWriter.writeNext(value.getHeaders());
+                                        for (String[] element : value.getData()) {
+                                            CsvWriter.writeNext(element);
+                                        }
+                                        CsvWriter.close();
+                                        fileData.setStatus(Status.SUCCESS);
+                                        fileStatusService.update(fileData);
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java]: Post the initial creation we are only updating the status. I think it will be better to have a method updateStatus(Long id, Status status) for this purpose

File: service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+                                        log.info("process complete");
+                                    }
+                                    catch (Exception exception)
+                                    {
+                                        log.error("retrying",exception);
+                                        throw exception;
+                                    }
+                                    return null;
+                                }, context -> {
+                                    log.error("retries exhausted");
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java]: You should log the actual exception as well

Comment (obedjamir) [reply service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java]: this has been implemented

File: service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java (hunk 15 above)
Comment (obedjamir) [reply service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java]: this has been implemented

File: service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java (hunk 14 above)
Comment (obedjamir) [reply service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java]: this has been removed

File: service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java (hunk 13 above)
Comment (obedjamir) [reply service/src/main/java/com/bankbazaar/kafka/service/config/KafkaStreamsConfig.java]: this method now return Data type
